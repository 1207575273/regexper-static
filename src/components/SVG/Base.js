import React from 'react';
import { Map } from 'immutable';

class Base extends React.PureComponent {
  _currentBBox() {
    return this.tempBBox ? this.tempBBox : (this.state || {}).bbox;
  }

  setStateAsync(state) {
    return new Promise(resolve => {
      this.setState(state, resolve);
    });
  }

  async setBBox(box, recalculate = {}) {
    const bbox = (this._currentBBox() || Map({ width: 0, height: 0})).withMutations(bbox => {
      bbox.merge(box);

      if (!bbox.has('axisY') || recalculate.axisY) {
        bbox.set('axisY', bbox.get('height') / 2);
      }

      if (!bbox.has('axisX1') || recalculate.axisX1) {
        bbox.set('axisX1', 0);
      }

      if (!bbox.has('axisX2') || recalculate.axisX2) {
        bbox.set('axisX2', bbox.get('width'));
      }
    });

    this.tempBBox = bbox; // Want to get the updated bbox while setState is pending
    await this.setStateAsync({ bbox });
    this.tempBBox = null;
  }

  getBBox() {
    const bbox = this._currentBBox() || Map();
    return {
      width: 0,
      height: 0,
      axisY: 0,
      axisX1: 0,
      axisX2: 0,
      ...bbox.toJS()
    };
  }

  async reflowChildren() {
    // No child components
    if (this.children === undefined) {
      return true;
    }

    const reflowed = await Promise.all(this.children.map(c => c.doReflow()));

    return reflowed.reduce((memo, value) => memo || value, false);
  }

  async doReflow() {
    const oldBBox = this._currentBBox();
    const shouldReflow = await this.reflowChildren();

    if (shouldReflow) {
      this.reflow();
    }

    return this._currentBBox() !== oldBBox;
  }

  reflow() {
    // Implemented in subclasses
  }
}

export default Base;
