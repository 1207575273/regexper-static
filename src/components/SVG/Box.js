import React from 'react';
import PropTypes from 'prop-types';

import Base from './Base';
import style from './style';

/** @extends React.PureComponent */
class Box extends Base {
  static defaultProps = {
    padding: 5,
    radius: 3
  }

  reflow() {
    const { padding, useAnchors } = this.props;
    const box = this.children[0].getBBox();
    const labelBox = this.label ? this.label.getBBox() : { width: 0, height: 0};

    this.setBBox({
      width: Math.max(box.width + 2 * padding, labelBox.width),
      height: box.height + 2 * padding + labelBox.height,
      axisY: (useAnchors ? box.axisY : box.height / 2) + padding + labelBox.height,
      axisX1: useAnchors ? box.axisX1 + padding : 0,
      axisX2: useAnchors ? box.axisX2 + padding : box.width + 2 * padding
    });

    return this.setStateAsync({
      width: this.getBBox().width,
      height: box.height + 2 * padding,
      contentTransform: `translate(${ padding } ${ padding + labelBox.height })`,
      rectTransform: `translate(0 ${ labelBox.height })`,
      labelTransform: `translate(0 ${ labelBox.height })`
    });
  }

  containedRef = contained => this.children = [contained]

  labelRef = label => this.label = label

  render() {
    const { theme, radius, label, children } = this.props;
    const { width, height, labelTransform, rectTransform, contentTransform } = this.state || {};

    const rectProps = {
      style: style[theme],
      width,
      height,
      rx: radius,
      ry: radius,
      transform: rectTransform
    };
    const textProps = {
      transform: labelTransform,
      style: style.infoText,
      ref: this.labelRef
    };

    return <React.Fragment>
      <rect { ...rectProps } ></rect>
      { label && <text { ...textProps }>{ label }</text> }
      <g transform={ contentTransform }>
        { React.cloneElement(React.Children.only(children), {
          ref: this.containedRef
        }) }
      </g>
    </React.Fragment>;
  }
}

Box.propTypes = {
  children: PropTypes.node,
  label: PropTypes.string,
  padding: PropTypes.number,
  useAnchors: PropTypes.bool,
  radius: PropTypes.number,
  theme: PropTypes.string
};

export default Box;
