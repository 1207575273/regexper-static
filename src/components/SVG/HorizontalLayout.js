import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';

import style from './style';

import reflowable from './reflowable';
import Path from './path';

@reflowable
class HorizontalLayout extends React.PureComponent {
  static defaultProps = {
    withConnectors: false,
    spacing: 10
  }

  state = {
    childTransforms: List()
  }

  updateChildTransforms(childBoxes) {
    return this.state.childTransforms.withMutations(transforms => (
      childBoxes.forEach((box, i) => (
        transforms.set(i, `translate(${ box.offsetX } ${ box.offsetY })`)
      ))
    ));
  }

  updateConnectorPaths(childBoxes) {
    let last = childBoxes[0];

    return childBoxes.slice(1).reduce((path, box) => {
      try {
        return path
          .moveTo({ x: last.offsetX + last.axisX2, y: this.getBBox().axisY })
          .lineTo({ x: box.offsetX + box.axisX1 });
      }
      finally {
        last = box;
      }
    }, new Path()).toString();
  }

  reflow() {
    const { spacing, withConnectors } = this.props;

    const childBoxes = this.children.map(child => child.getBBox());
    const verticalCenter = childBoxes.reduce((center, box) => Math.max(center, box.axisY), 0);
    const width = childBoxes.reduce((width, box) => width + box.width, 0) + (childBoxes.length - 1) * spacing;
    const height = childBoxes.reduce((ascHeight, box) => Math.max(ascHeight, box.axisY), 0) +
                   childBoxes.reduce((decHeight, box) => Math.max(decHeight, box.height - box.axisY), 0);
    this.setBBox({ width, height, axisY: verticalCenter }, { axisX1: true, axisX2: true });

    let offset = 0;
    childBoxes.forEach(box => {
      box.offsetX = offset;
      box.offsetY = this.getBBox().axisY - box.axisY;
      offset += box.width + spacing;
    });

    this.setStateAsync({
      childTransforms: this.updateChildTransforms(childBoxes),
      connectorPaths: withConnectors ? this.updateConnectorPaths(childBoxes) : ''
    });
  }

  childRef = i => child => this.children[i] = child

  render() {
    const { children } = this.props;
    const { childTransforms, connectorPaths } = this.state;

    this.children = [];

    return <React.Fragment>
      <path d={ connectorPaths } style={ style.connectors }></path>
      { React.Children.map(children, (child, i) => (
        <g transform={ childTransforms.get(i) }>
          { React.cloneElement(child, {
            ref: this.childRef(i)
          }) }
        </g>
      ))}
    </React.Fragment>;
  }
}

HorizontalLayout.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
  spacing: PropTypes.number,
  withConnectors: PropTypes.bool
};

export default HorizontalLayout;
