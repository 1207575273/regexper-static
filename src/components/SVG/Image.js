import React from 'react';
import PropTypes from 'prop-types';

import style from './style';

import reflowable from './reflowable';

const namespaceProps = {
  'xmlns': 'http://www.w3.org/2000/svg',
  'xmlns:cc': 'http://creativecommons.org/ns#',
  'xmlns:rdf': 'http://www.w3.org/1999/02/22-rdf-syntax-ns#'
};
const metadata = `<rdf:rdf>
 <cc:license rdf:about="http://creativecommons.org/licenses/by/3.0/">
    <cc:permits rdf:resource="http://creativecommons.org/ns#Reproduction"></cc:permits>
    <cc:permits rdf:resource="http://creativecommons.org/ns#Distribution"></cc:permits>
    <cc:requires rdf:resource="http://creativecommons.org/ns#Notice"></cc:requires>
    <cc:requires rdf:resource="http://creativecommons.org/ns#Attribution"></cc:requires>
    <cc:permits rdf:resource="http://creativecommons.org/ns#DerivativeWorks"></cc:permits>
  </cc:license>
</rdf:rdf>`;

@reflowable
class Image extends React.PureComponent {
  static defaultProps = {
    padding: 10
  }

  state = {
    width: 0,
    height: 0
  }

  reflow() {
    const { padding } = this.props;
    const box = this.children[0].getBBox();

    return this.setStateAsync({
      width: Math.round(box.width + 2 * padding),
      height: Math.round(box.height + 2 * padding)
    });
  }

  containedRef = contained => this.children = [contained]

  svgRef = svg => this.svg = svg

  render() {
    const { width, height } = this.state;
    const { padding, children } = this.props;

    const svgProps = {
      width,
      height,
      viewBox: [0, 0, width, height].join(' '),
      style: style.image,
      ref: this.svgRef,
      ...namespaceProps
    };

    return <svg { ...svgProps }>
      <metadata dangerouslySetInnerHTML={{ __html: metadata }}></metadata>
      <g transform={ `translate(${ padding } ${ padding })` }>
        { React.cloneElement(React.Children.only(children), {
          ref: this.containedRef
        }) }
      </g>
    </svg>;
  }
}

Image.propTypes = {
  children: PropTypes.node,
  padding: PropTypes.number
};

export default Image;
