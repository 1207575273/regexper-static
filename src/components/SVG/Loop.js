import React from 'react';
import PropTypes from 'prop-types';

import Base from './Base';
import style from './style';
import Path from './path';

const skipPath = (box, greedy) => {
  const vert = Math.max(0, box.axisY - 10);
  const horiz = box.width - 10;

  let path = new Path({ relative: true });

  if (!greedy) {
    path
      .moveTo({ x: 10, y: box.axisY + box.offsetY - 15, relative: false })
      .lineTo({ x: 5, y: 5 })
      .moveTo({ x: -5, y: -5 })
      .lineTo({ x: -5, y: 5 });
  }

  return path
    .moveTo({ x: 0, y: box.axisY + box.offsetY, relative: false })
    .quadraticCurveTo({ cx: 10, cy: 0, x: 10, y: -10 })
    .lineTo({ y: -vert })
    .quadraticCurveTo({ cx: 0, cy: -10, x: 10, y: -10 })
    .lineTo({ x: horiz })
    .quadraticCurveTo({ cx: 10, cy: 0, x: 10, y: 10 })
    .lineTo({ y: vert })
    .quadraticCurveTo({ cx: 0, cy: 10, x: 10, y: 10 });
};

const repeatPath = (box, greedy) => {
  const vert = box.height - box.axisY - 10;

  let path = new Path({ relative: true });

  if (greedy) {
    path
      .moveTo({ x: box.offsetX + box.width + 10, y: box.axisY + box.offsetY + 15, relative: false })
      .lineTo({ x: 5, y: -5 })
      .moveTo({ x: -5, y: 5 })
      .lineTo({ x: -5, y: -5 });
  }

  return path
    .moveTo({ x: box.offsetX, y: box.axisY + box.offsetY, relative: false })
    .quadraticCurveTo({ cx: -10, cy: 0, x: -10, y: 10 })
    .lineTo({ y: vert })
    .quadraticCurveTo({ cx: 0, cy: 10, x: 10, y: 10 })
    .lineTo({ x: box.width })
    .quadraticCurveTo({ cx: 10, cy: 0, x: 10, y: -10 })
    .lineTo({ y: -vert })
    .quadraticCurveTo({ cx: 0, cy: -10, x: -10, y: -10 });
};

/** @extends React.PureComponent */
class Loop extends Base {
  get contentOffset() {
    const { skip, repeat } = this.props;

    if (skip) {
      return { x: 15, y: 10 };
    } else if (repeat) {
      return { x: 10, y: 0 };
    } else {
      return { x: 0, y: 0 };
    }
  }

  preReflow() {
    return this.contained;
  }

  reflow() {
    return new Promise(resolve => {
      const { skip, repeat, greedy } = this.props;
      const box = this.contained.getBBox();
      const labelBox = this.label ? this.label.getBBox() : { width: 0, height: 0 };

      let height = box.height + labelBox.height;
      if (skip) {
        height += 10;
      }
      if (repeat) {
        height += 10;
      }

      this.setBBox({
        width: box.width + this.contentOffset.x * 2,
        height,
        axisY: box.axisY + this.contentOffset.y,
        axisX1: box.axisX1 + this.contentOffset.x,
        axisX2: box.axisX2 + this.contentOffset.x
      });

      box.offsetX = this.contentOffset.x;
      box.offsetY = this.contentOffset.y;

      this.setState({
        labelTransform: `translate(${ this.getBBox().width - labelBox.width - 10 } ${ this.getBBox().height + 2 })`,
        loopPaths: [
          skip && skipPath(box, greedy),
          repeat && repeatPath(box, greedy)
        ].filter(Boolean).join('')
      }, resolve);
    });
  }

  containedRef = contained => this.contained = contained

  labelRef = label => this.label = label

  render() {
    const { label, children } = this.props;
    const { loopPaths, labelTransform } = this.state || {};

    const textProps = {
      transform: labelTransform,
      style: style.infoText,
      ref: this.labelRef
    };

    return <React.Fragment>
      <path d={ loopPaths } style={ style.connectors }></path>
      { label && <text { ...textProps }>{ label }</text> }
      <g transform={ `translate(${ this.contentOffset.x } ${ this.contentOffset.y })` }>
        { React.cloneElement(React.Children.only(children), {
          ref: this.containedRef
        }) }
      </g>
    </React.Fragment>;
  }
}

Loop.propTypes = {
  children: PropTypes.node.isRequired,
  greedy: PropTypes.bool,
  label: PropTypes.string,
  skip: PropTypes.bool,
  repeat: PropTypes.bool
};

export default Loop;
