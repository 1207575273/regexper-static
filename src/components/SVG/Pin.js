import React from 'react';
import PropTypes from 'prop-types';

import style from './style';

import reflowable from './reflowable';

@reflowable
class Pin extends React.PureComponent {
  static defaultProps = {
    radius: 5
  }

  reflow() {
    const { radius } = this.props;

    this.setBBox({
      width: radius * 2,
      height: radius * 2
    });
  }

  render() {
    const { radius } = this.props;

    const circleProps = {
      r: radius,
      style: style.pin,
      transform: `translate(${ radius } ${ radius })`
    };

    return <circle { ...circleProps }></circle>;
  }
}

Pin.propTypes = {
  radius: PropTypes.number
};

export default Pin;
