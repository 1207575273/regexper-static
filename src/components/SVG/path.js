class Path {
  constructor({ relative } = {}) {
    this.relative = relative || false;
    this.currentPosition = { x: 0, y: 0 };
    this.lastStartPosition = { x: 0, y: 0 };
    this.pathParts = [];
  }

  moveTo({ x, y, relative }) {
    relative = relative === undefined ? this.relative : relative;
    const command = relative ? 'm' : 'M';

    if (relative) {
      this.currentPosition = {
        x: x + this.currentPosition.x,
        y: y + this.currentPosition.y
      };
    } else {
      this.currentPosition = { x, y };
    }

    this.pathParts.push(`${command}${x},${y}`);
    this.lastStartPosition = { ...this.currentPosition };

    return this;
  }

  lineTo({ x, y, relative }) {
    if (relative === undefined) {
      relative = this.relative;
    }
    if (x === undefined) {
      x = relative ? 0 : this.currentPosition.x;
    }
    if (y === undefined) {
      y = relative ? 0 : this.currentPosition.y;
    }

    if ((relative && x === 0 && y === 0)
      || (!relative && x === this.currentPosition.x && y === this.currentPosition.y)) {
      return this;
    }

    if (y === (relative ? 0 : this.currentPosition.y)) {
      const command = relative ? 'h' : 'H';
      this.pathParts.push(`${command}${x}`);
    } else if (x === (relative ? 0 : this.currentPosition.x)) {
      const command = relative ? 'v' : 'V';
      this.pathParts.push(`${command}${y}`);
    } else {
      const command = relative ? 'l' : 'L';
      this.pathParts.push(`${command}${x},${y}`);
    }

    this.currentPosition = {
      x: relative ? this.currentPosition.x + x : x,
      y: relative ? this.currentPosition.y + y : y
    };

    return this;
  }

  closePath() {
    this.pathParts.push('Z');
    this.currentPosition = { ...this.lastStartPosition };
    return this;
  }

  cubicCurveTo({ cx1, cy1, cx2, cy2, x, y, relative }) {
    relative = relative === undefined ? this.relative : relative;
    if (cx1 === undefined || cy1 === undefined) {
      const command = relative ? 's' : 'S';
      this.pathParts.push(`${command}${cx2},${cy2} ${x},${y}`);
    } else {
      const command = relative ? 'c' : 'C';
      this.pathParts.push(`${command}${cx1},${cy1} ${cx2},${cy2} ${x},${y}`);
    }

    this.currentPosition.x = relative ? this.currentPosition.x + x : x;
    this.currentPosition.y = relative ? this.currentPosition.y + y : y;

    return this;
  }

  quadraticCurveTo({ cx, cy, x, y, relative }) {
    relative = relative === undefined ? this.relative : relative;
    if (cx === undefined || cy === undefined) {
      const command = relative ? 't' : 'T';
      this.pathParts.push(`${command} ${x},${y}`);
    } else {
      const command = relative ? 'q' : 'Q';
      this.pathParts.push(`${command}${cx},${cy} ${x},${y}`);
    }

    this.currentPosition.x = relative ? this.currentPosition.x + x : x;
    this.currentPosition.y = relative ? this.currentPosition.y + y : y;

    return this;
  }

  arcTo({ rx, ry, rotation, arc, sweep, x, y, relative }) {
    relative = relative === undefined ? this.relative : relative;
    const command = relative ? 'a' : 'A';

    this.pathParts.push(`${command}${rx},${ry} ${rotation} ${arc} ${sweep},${x},${y}`);

    this.currentPosition.x = relative ? this.currentPosition.x + x : x;
    this.currentPosition.y = relative ? this.currentPosition.y + y : y;

    return this;
  }

  toString() {
    return this.pathParts.join('');
  }
}

export default Path;
