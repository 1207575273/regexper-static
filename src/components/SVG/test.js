import React from 'react';
import { shallow } from 'enzyme';

import SVG from 'components/SVG';

describe('SVG', () => {
  test('rendering', async () => {
    const data = {
      type: 'Pin'
    };
    const component = shallow(
      <SVG data={ data }/>
    );
    expect(component).toMatchSnapshot();
  });

  test('rendering 1 child', () => {
    const data = {
      type: 'Image',
      children: [
        {
          type: 'Pin'
        }
      ]
    };
    const component = shallow(
      <SVG data={ data }/>
    );
    expect(component).toMatchSnapshot();
  });

  test('rendering multiple children', () => {
    const data = {
      type: 'HorizontalLayout',
      children: [
        {
          type: 'Pin'
        },
        {
          type: 'Pin'
        }
      ]
    };
    const component = shallow(
      <SVG data={ data }/>
    );
    expect(component).toMatchSnapshot();
  });

  test('rendering a string as a child', () => {
    const data = {
      type: 'Text',
      children: [
        'Sample text'
      ]
    };
    const component = shallow(
      <SVG data={ data }/>
    );
    expect(component).toMatchSnapshot();
  });

  test('passing props to nodes', () => {
    const data = {
      type: 'Text',
      props: {
        theme: 'anchorText'
      },
      children: [
        'Sample text'
      ]
    };
    const component = shallow(
      <SVG data={ data }/>
    );
    expect(component).toMatchSnapshot();
  });

  test('arbitrary markup', () => {
    const data = {
      type: 'Text',
      children: [
        {
          type: 'tspan',
          children: [
            'Sample text'
          ]
        }
      ]
    };
    const component = shallow(
      <SVG data={ data }/>
    );
    expect(component).toMatchSnapshot();
  });
});
