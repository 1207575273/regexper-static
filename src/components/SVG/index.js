import React from 'react';

import Box from './Box';
import HorizontalLayout from './HorizontalLayout';
import Image from './Image';
import Loop from './Loop';
import Pin from './Pin';
import Text from './Text';
import VerticalLayout from './VerticalLayout';

const nodeTypes = {
  Box,
  HorizontalLayout,
  Image,
  Loop,
  Pin,
  Text,
  VerticalLayout
};

const renderChildren = children => {
  if (!children || children.length === 0) {
    return;
  }

  return children.length === 1 ?
    renderImage(children[0]) :
    children.map((node, i) => renderImage(node, { key: i }));
};

const renderImage = (node, extraProps = {}) => {
  if (typeof node === 'string') {
    return node;
  }

  const { type, props, children } = node;

  return React.createElement(nodeTypes[type], { ...extraProps, ...props }, renderChildren(children));
};

export default renderImage;
