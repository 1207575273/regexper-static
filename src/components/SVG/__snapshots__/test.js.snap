// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVG arbitrary markup 1`] = `
<Text>
  <tspan
    key="0"
  >
    Sample text
  </tspan>
</Text>
`;

exports[`SVG passing props to nodes 1`] = `
<Text
  theme="anchorText"
>
  Sample text
</Text>
`;

exports[`SVG rendering 1`] = `
<Pin
  radius={5}
/>
`;

exports[`SVG rendering 1 child 1`] = `
<Image
  padding={10}
>
  <Pin
    key="0"
    radius={5}
  />
</Image>
`;

exports[`SVG rendering a string as a child 1`] = `
<Text>
  Sample text
</Text>
`;

exports[`SVG rendering multiple children 1`] = `
<HorizontalLayout
  spacing={10}
  withConnectors={false}
>
  <Pin
    key="0"
    radius={5}
  />
  <Pin
    key="1"
    radius={5}
  />
</HorizontalLayout>
`;
